pkg/products/threescale/reconciler.go

```diff
@@ -1340,35 +1343,147 @@ func (r *Reconciler) updateKeycloakUsersAttributeWith3ScaleUserId(ctx context.Co
+					Namespace: r.Config.GetNamespace(),
+				},
+			}
+			err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: signUpAccountsSecret.Name, Namespace: signUpAccountsSecret.Namespace}, signUpAccountsSecret)
+			if err != nil {
+				return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("Error getting access token secret: %w", err)
+			}
+
+			if val, ok := signUpAccountsSecret.Data[string(account.OrgName)]; ok {
+				signUpAccount := SignUpAccount{
+					AccountDetail: account,
+					AccountAccessToken: AccountAccessToken{
+						Value: string(val),
+					},
+				}
+				// verify if the account have the auth provider already
+				err = r.AddAuthProviderToMTAccount(ctx, serverClient, signUpAccount)
+				r.log.Infof("Add auth provider for missing account:",
+					l.Fields{"accountsWithProvider": signUpAccount},
+				)
+				if err != nil {
+					return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("Error adding auth provider: %w", err)
+				}
+			}
+			err = r.reconcileDashboardLink(ctx, serverClient, account.OrgName, account.AdminBaseURL)
+			if err != nil {
+				return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("error reconciling console link: %v", err)
+			}
+
+		}
+	}
 
 	// creating new MT accounts in 3scale
-	accountsToBeCreated := getMTAccountsToBeCreated(identities.Items, accounts)
-	r.log.Infof("Creating new MT accounts:", l.Fields{"accountsToBeCreated": accountsToBeCreated})
-	r.tsClient.CreateTenants(*accessToken, accountsToBeCreated)
-	if err != nil {
-		r.log.Error("Error creating new tenant accounts:", err)
-		return integreatlyv1alpha1.PhaseFailed, err
+	accountsToBeCreated := getMTAccountsToBeCreated(mtUserIdentities.Items, accounts)
+	r.log.Infof("retrieving new MT accounts to be created",
+		l.Fields{"accountsToBeCreated": accountsToBeCreated},
+	)
+
+	// TODO: create CM with account state of each account
+
+	for _, account := range accountsToBeCreated {
+		// create account
+		newSignupAccount, err := r.tsClient.CreateTenant(*accessToken, account)
+		if err != nil {
+			r.log.Error("Error creating new tenant accounts:", err)
+			return integreatlyv1alpha1.PhaseFailed, err
+		}
+		r.log.Infof("new MT account created",
+			l.Fields{"signupAccount": newSignupAccount},
+		)
+		// TODO: store token in a secret
+		signUpAccountsSecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "mt-signupaccount-3scale-access-token",
+				Namespace: r.Config.GetNamespace(),
+			},
+		}
+
+		err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: signUpAccountsSecret.Name, Namespace: signUpAccountsSecret.Namespace}, signUpAccountsSecret)
+		if !k8serr.IsNotFound(err) && err != nil {
+			return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("Error getting access token secret: %w", err)
+		} else if k8serr.IsNotFound(err) {
+			signUpAccountsSecret.Data = map[string][]byte{}
+		}
+
+		signUpAccountsSecret.ObjectMeta.ResourceVersion = ""
+		signUpAccountsSecret.Data[string(account.OrgName)] = []byte(newSignupAccount.AccountAccessToken.Value)
```

Comment:
Could data be nil here?

Reply:
I don't think so, there is an if statement above that checks whether the secret exists or not and if not it assigns a map to Data

https://github.com/integr8ly/integreatly-operator/pull/2016/files#diff-c99ff6dd3de866743c817f7b1dc230b4875e42eba21e5768f34fd92b9f873e32R1451